### PR TITLE
 React to APICompat changes enabling reverse APICompat for facades

### PR DESCRIPTION
**Pull request by ericstj**
I made changes to API Compat that enable it for partial facades, this is the result.  We'll want to look through these baselines before RTM to understand if anything here needs to be exposed.

For things that need to be exposed that can happen in separate PRs since it may require API review + test additions. 

NO MERGE - this is using a private build of API compat.

### CODE REVIEW

**Comment by ericstj** [comment]
@ViktorHofer @safern : tests tab shows no failures, yet the send to helix step is reporting a failure.  Might be something amiss in the AZDO test integration.

**Comment by ViktorHofer** [comment]
Can you point me to the AzDO build? Usually that happens when the work item starts to fail. We mentioned that in the mail that I sent out earlier this week.

**Comment by ericstj** [comment]
https://dev.azure.com/dnceng/public/_build/results?buildId=177761&view=logs&jobId=29e7c093-c33b-5b8e-a00b-0e954875c839&taskId=e8d9965f-ee57-59e4-3e7a-0c8aaa58218b&lineStart=42&lineEnd=44&colStart=1&colEnd=14

**Comment by ViktorHofer** [comment]
https://github.com/dotnet/arcade/issues/2659#issuecomment-489123733

File: src/System.Utf8String.Experimental/src/MatchingRefApiCompatBaseline.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Compat issues with assembly System.Utf8String.Experimental:
+MembersMustExist : Member 'System.Utf8String.Substring(System.Index)' does not exist in the reference but it does exist in the implementation.
```
Comment (ahsonkhan) [inline src/System.Utf8String.Experimental/src/MatchingRefApiCompatBaseline.txt]: @GrabYourPitchforks, this should/can be fixed in the ref.

https://github.com/dotnet/coreclr/blob/4ad11ccba4ebb458046891bd07c0c7d4440d7df4/src/System.Private.CoreLib/src/System/Utf8String.Manipulation.cs#L30-L35

Comment (ahsonkhan) [reply src/System.Utf8String.Experimental/src/MatchingRefApiCompatBaseline.txt]: Unless it was meant to be removed as part of https://github.com/dotnet/coreclr/pull/24036. cc @tarekgh

Comment (tarekgh) [reply src/System.Utf8String.Experimental/src/MatchingRefApiCompatBaseline.txt]: @ahsonkhan yes this method is removed. callers can use Utf8String[Index.GetOffset(Length)..] instead.

Comment (ericstj) [reply src/System.Utf8String.Experimental/src/MatchingRefApiCompatBaseline.txt]: It is still present in the implementation, thus the baseline. https://github.com/dotnet/coreclr/blob/c2e71fd5e711ca150e238741015a1a0229606793/src/System.Private.CoreLib/src/System/Utf8String.Manipulation.cs#L31

**Comment by ericstj** [comment]
Removing NO MERGE.  This is now ready to go.   FYI: this means we now will run reverse API compat on all implmentation assemblies (including partial facades) so it may mean more things to fix in the coreclr ingestion PRs.

**Comment by ericstj** [comment]
@ViktorHofer looks like this might be failing due to https://github.com/dotnet/arcade/pull/2402

**Comment by ViktorHofer** [comment]
Oh right. Let me fix that.

**Comment by ViktorHofer** [comment]
Fixed, the host was missing. This is a workaround, I would have expected the build file to deal with copying the host to the output dir. Seems like depproj files don't import such.

**Comment by ericstj** [comment]
> Fixed, the host was missing. This is a workaround, I would have expected the build targets file to deal with copying the host to the output dir. Seems like depproj files don't import such.

That's what I thought.  It seemed to me like restore didn't even write this to the generated targets file, so it might be an issue with the naming convention in the package.  Here's what I see in the assets file: 
```json
  "targets": {
    ".NETCoreApp,Version=v3.0": {
...
      "Microsoft.DotNet.RemoteExecutor/1.0.0-beta.19256.12": {
        "type": "package",
        "dependencies": {
          "xunit.assert": "2.4.1-pre.build.4059"
        },
        "compile": {
          "lib/netcoreapp2.1/Microsoft.DotNet.RemoteExecutor.dll": {}
        },
        "runtime": {
          "lib/netcoreapp2.1/Microsoft.DotNet.RemoteExecutor.dll": {}
        },
        "build": {
          "build/_._": {}
        }
      },
```

Maybe something is adding ExcludeAssets?

File: eng/resolveContract.targets [hunk 2]
```diff
@@ -2,7 +2,10 @@
   <PropertyGroup Condition="'$(IsSourceProject)' == 'true'">
     <ContractProject Condition="'$(ContractProject)' == ''">$(SourceDir)/$(MSBuildProjectName)/ref/$(MSBuildProjectName).csproj</ContractProject>
     <HasMatchingContract Condition="'$(HasMatchingContract)' == '' and Exists('$(ContractProject)')">true</HasMatchingContract>
-    <ContractAssemblyPath Condition="'$(ContractAssemblyPath)' == ''">$(ContractOutputPath)/$(MSBuildProjectName).dll</ContractAssemblyPath>
+    <ContractAssemblyPath Condition="'$(ContractAssemblyPath)' == ''">$(RefPath)/$(MSBuildProjectName).dll</ContractAssemblyPath>
+
+    <!-- Disable API compat if the project doesn't have reference assembly -->
+    <RunApiCompat Condition="'$(HasMatchingContract)' != 'true'">false</RunApiCompat>
```
Comment (joperezr) [inline eng/resolveContract.targets]: Won't you also have to set `RunMatchingRefApiCompat` to false in this case? or is this property a switch for both src and reverse api compat?

File: eng/resolveContract.targets [hunk 3]
```diff
@@ -17,7 +20,13 @@
   <!-- intentionally empty since we no longer need a target -->
   <Target Name="ResolveMatchingContract" />
 
-  <Target Name="VerifyMatchingContract" AfterTargets="ResolveReferences">
-    <Error Condition="'$(HasMatchingContract)' == 'true' and !Exists('%(ResolvedMatchingContract.Identity)')" Text="ResolveMatchingContract could not find a matching contract '%(ResolvedMatchingContract.Identity)' not found." />
+  <!-- Disable matching ref API compat for aot runtime refencing libs, these have a lot of public API
+         only by implementation and we don't intend to start new API development on AOT -->
+  <Target Name="DisableRunMatchingRefApiCompat"
```
Comment (joperezr) [inline eng/resolveContract.targets]: Just wondering, why does this have to happen via a Target and not defined statically like for non-aot libraries.

File: external/test-runtime/XUnit.Runtime.depproj [hunk 4]
```diff
@@ -309,4 +309,25 @@
 
   </Target>
 
+  <Target Name="AddRemoteExecutorHostTools"
```
Comment (joperezr) [inline external/test-runtime/XUnit.Runtime.depproj]: I understand we would want this change but I'm not sure how is this related to ApiCompat changes. I assume this is just needed because of other Arcade changes?

File: src/CoreFx.Private.TestUtilities/src/CoreFx.Private.TestUtilities.csproj [hunk 5]
```diff
@@ -8,6 +8,7 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <NoWarn>$(NoWarn);CS1573</NoWarn>
     <EnablePInvokeAnalyzer>false</EnablePInvokeAnalyzer>
+    <RunMatchingRefApiCompat>false</RunMatchingRefApiCompat>
```
Comment (joperezr) [inline src/CoreFx.Private.TestUtilities/src/CoreFx.Private.TestUtilities.csproj]: NIT: Should we add a comment here saying why we want to exclude the reverse apicompat run? I only ask because it seems like if we did we would have catched the API you are adding right now on the TestUtitlities.cs ref code.

**Comment by joperezr** [comment]
NIT: looks like you ran a msbuild /bl and checked in the msbuild.ProjectImports.zip by mistake on the AsyncInterfaces src directory 😆

File: src/System.Runtime.Serialization.Formatters/src/MatchingRefApiCompatBaseline.txt [hunk 6]
```diff
@@ -1,3 +1,9 @@
 # Exposed publicly only in implementation for serialization compat
 TypesMustExist : Type 'System.Runtime.Serialization.SerializationEventHandler' does not exist in the reference but it does exist in the implementation.
 TypesMustExist : Type 'System.Runtime.Serialization.TypeLoadExceptionHolder' does not exist in the reference but it does exist in the implementation.
+
+MembersMustExist : Member 'System.Runtime.Serialization.SerializationInfo.DeserializationInProgress.get()' does not exist in the reference but it does exist in the implementation.
```
Comment (joperezr) [inline src/System.Runtime.Serialization.Formatters/src/MatchingRefApiCompatBaseline.txt]: NIT: Looks like this baseline has an extra space before the first MembersMustExist.

File: eng/resolveContract.targets (hunk 2 above)
Comment (ericstj) [reply eng/resolveContract.targets]: RunApiCompat controls both directions.

File: external/test-runtime/XUnit.Runtime.depproj (hunk 4 above)
Comment (ericstj) [reply external/test-runtime/XUnit.Runtime.depproj]: I deleted it.  It was reacting to other arcade changes.  See @ViktorHofer and my comments above.

File: eng/resolveContract.targets (hunk 3 above)
Comment (ericstj) [reply eng/resolveContract.targets]: I wanted to match the condition that was previously used where this is disabled when `@(ReferenceFromRuntime)` is used.  Since ReferenceFromRuntime is an item and we need to set a property based on its presence it must be done in a Target.  I could have conditioned on "IsPartialFacade" instead but that would disable APICompat in more cases than are necessary.  I could have just ignored this and instead added baselines as we did for non-aot.  I decided not to do so because right now I'm trying to make AOT less of a tax on this repo until we decide how were going to unify it with the 3.0 codebase.

File: src/CoreFx.Private.TestUtilities/src/CoreFx.Private.TestUtilities.csproj (hunk 5 above)
Comment (ericstj) [reply src/CoreFx.Private.TestUtilities/src/CoreFx.Private.TestUtilities.csproj]: It doesn't consistently expose API across all its configurations.  I tried to fix it in one case but found I couldn't fix it everywhere without adding baselines or removing API from some frameworks.  Since we don't care so much about this thing being used externally I felt that API not being present for tests to consume would be a good enough guard against folks forgetting to expose it.

File: src/System.Runtime.Serialization.Formatters/src/MatchingRefApiCompatBaseline.txt (hunk 6 above)
Comment (ericstj) [reply src/System.Runtime.Serialization.Formatters/src/MatchingRefApiCompatBaseline.txt]: intentional since the comment above doesn't necessarily apply to the ones that I added.

**Comment by ericstj** [comment]
> build errors looks like you are still missing a baseline in System.Threading.Tasks.

Yeah, this is do to @stephentoub's rename.  We can baseline them until we get a build from CoreCLR with them removed.

File: src/System.Runtime.Intrinsics/src/MatchingRefApiCompatBaseline.txt [hunk 7]
```diff
@@ -0,0 +1,17 @@
+Compat issues with assembly System.Runtime.Intrinsics:
+CannotRemoveAttribute : Attribute 'System.CLSCompliantAttribute' exists on 'System.Runtime.Intrinsics.X86.Aes' in the implementation but not the reference.
```
Comment (tannergooding) [inline src/System.Runtime.Intrinsics/src/MatchingRefApiCompatBaseline.txt]: Is `CLSCompliantAttribute` meaningful? If so, we should just add it...

Comment (ericstj) [reply src/System.Runtime.Intrinsics/src/MatchingRefApiCompatBaseline.txt]: It needs to be consistent.  I believe the problem is that CoreLib sets CLSCompliant on assembly level then turns if off on individual types.  Your reference assembly sets it as false on the assembly so it doesn't need to disable it on individual types. https://github.com/dotnet/corefx/blob/d81fc3589a166ab78096dee0a4ddf0cd2fbd17d7/src/System.Runtime.Intrinsics/ref/System.Runtime.Intrinsics.csproj#L4

**Comment by stephentoub** [comment]
> Yeah, this is do to @stephentoub's rename from this morning

Sorry, did I somehow break something in master?

**Comment by ericstj** [comment]
@stephentoub nope, just broke this PR between the time I tested locally and when the CI hit it.  No worries.

@ViktorHofer your test host change is still broken.  The problem is that it's referenced by test projects via the `$(RuntimePath)` which will have netcoreapp assemblies.  These are incorrect when building a `netstandard` test project.  For now I'm going to undo the update to the remote executor host so that you can work on a fix for that.

**Comment by ViktorHofer** [comment]
>  These are incorrect when building a netstandard test project.

That surprises me as the Remote Executor lib assembly is binary identical even though it cross-compiles.

**Comment by ericstj** [comment]
> That surprises me as the Remote Executor lib assembly is binary identical even though it cross-compiles.

The references are different.  .NETCoreApp is System.Runtime based and the System.Runtime assembly version is higher than the facades included in netstandard projects.

**Comment by ericstj** [comment]
All tests are failing with:
```
error: Common Language Runtime detected an invalid program.
```

This appears to be a result of arcade changes, but I haven't isolated the change.  I'm going to dial back this PR to only the API compat update.  We'll need to further investigate what's causing this, but we can do so in the Arcade ingestion PR.

**Comment by ericstj** [comment]
@jkoritzinsky it looks like your change https://github.com/dotnet/coreclr/commit/796776c2c340947fd092ccb9ddab579c41c0fd77 is causing CoreFx tests to fail when run on nano.  It appear that this is due to https://github.com/dotnet/arcade/blob/ce8e852e418ca52d28e562f9697fbb926d0b0bb1/src/Microsoft.DotNet.XUnitConsoleRunner/src/Program.cs#L8.

**Comment by ericstj** [comment]
Only failure is network test.
